c/experimental/stf/include/cccl/c/experimental/stf/stf.h

```diff
@@ -0,0 +1,224 @@
+#include <assert.h>
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+// TODO use CCCL_C_EXTERN_C_BEGIN/CCCL_C_EXTERN_C_END
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum stf_access_mode
+{
+  STF_NONE  = 0,
+  STF_READ  = 1 << 0,
+  STF_WRITE = 1 << 1,
+  STF_RW    = STF_READ | STF_WRITE
+} stf_access_mode;
+
+struct stf_exec_place_device
+{
+  int dev_id;
+};
+
+struct stf_exec_place_host
+{
+  char dummy; /* dummy to keep it standard C which does not allow empty structs */
+};
+
+typedef enum stf_exec_place_kind
+{
+  STF_EXEC_PLACE_DEVICE,
+  STF_EXEC_PLACE_HOST
+} stf_exec_place_kind;
+
+struct stf_exec_place
+{
+  enum stf_exec_place_kind kind;
+  union
+  {
+    struct stf_exec_place_device device;
+    struct stf_exec_place_host host;
+  } u;
+};
+
+static inline struct stf_exec_place make_device_place(int dev_id)
+{
+  struct stf_exec_place p;
+  p.kind            = STF_EXEC_PLACE_DEVICE;
+  p.u.device.dev_id = dev_id;
+  return p;
+}
+
+static inline struct stf_exec_place make_host_place()
+{
+  struct stf_exec_place p;
+  p.kind         = STF_EXEC_PLACE_HOST;
+  p.u.host.dummy = 0; /* to avoid uninitialized memory warnings */
+  return p;
+}
+
+typedef struct stf_exec_place_device stf_exec_place_device;
+typedef struct stf_exec_place_host stf_exec_place_host;
+typedef union stf_exec_place_u stf_exec_place_u;
+typedef struct stf_exec_place stf_exec_place;
```

Comment:
For consistency with existing code, could we move these to the struct definition?